### PR TITLE
Add support for publishing macOS M1 ARM64 wheels

### LINKED ISSUE #2503: tensorflow-addons doesn't work with tensorflow-macos
**Describe the feature and the current behavior/state.**
On my Mac with M1, on which [TensorFlow 2.5 for Mac](https://developer.apple.com/metal/tensorflow-plugin/) is installed, `pip install tensorflow-addons` fails with this error:

```
% pip install tensorflow-addons
ERROR: Could not find a version that satisfies the requirement tensorflow-addons (from versions: none)
ERROR: No matching distribution found for tensorflow-addons
```

Here's the output of `pip list`:
```
% pip list
Package                 Version
----------------------- -------------------
...
tensorboard-plugin-wit  1.8.0
tensorflow-estimator    2.5.0
tensorflow-macos        2.5.0 <--
tensorflow-metal        0.1.1
termcolor               1.1.0
...
```

**Relevant information**
- Are you willing to contribute it (yes/no): maybe
*If you wish to contribute, then read the requirements for new contributions in [`CONTRIBUTING.md`](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#requirements-for-new-contributions-to-the-repository)*
- Are you willing to maintain it going forward? (yes/no):
- Is there a relevant academic paper? (if so, where): no
- Does the relavent academic paper exceed 50 citations? (yes/no): n/a
- Is there already an implementation in another framework? (if so, where):
- Was it part of tf.contrib? (if so, where): setup.py

**Who will benefit with this feature?**
Users of TensorFlow for Mac.

**Any other info.**

### PR DESCRIPTION
**Pull request by lgeiger**
# Description

This PR adds a release script and updated the GitHub actions workflow to publish cross compiled ARM64 macOS wheels. Recent XCode version support cross compilation, so we can generate ARM64 wheels on the x86 GitHub actions VMs.

Fixes #2503

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [x] By running Black + Flake8
    - [x] By running pre-commit hooks
- [x] This PR addresses an already submitted issue for TensorFlow Addons
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

I manually verified the generated wheel on a ARM64 M1 Mac.

~~I am not sure what to do about the `Run the cpu tests in a small python docker image` CI failure. It seems unrelated to this PR.  @seanpmorgan What do you think?~~

### CODE REVIEW

File: .github/workflows/make_wheel_macOS_arm64.sh
Comment (lgeiger) [inline .github/workflows/make_wheel_macOS_arm64.sh:11]: This CPU setting makes sure that we cross compile for ARM64.

File: .github/workflows/make_wheel_macOS_arm64.sh
Comment (lgeiger) [inline .github/workflows/make_wheel_macOS_arm64.sh:13]: We don't need to support 10.13 since the only the latest version macOS version currently supports ARM64. But setting a minimum version here makes sure that the script will keep working even if the GitHub actions VM are updated to 11.1 at some point in the future.

File: .github/workflows/make_wheel_macOS_arm64.sh
Comment (lgeiger) [inline .github/workflows/make_wheel_macOS_arm64.sh:20]: We need to manually set the correct wheel name, since we are cross compiling on an x86 machine.

File: .github/workflows/release.yml
Comment (lgeiger) [inline .github/workflows/release.yml:53]: [`tensorflow-macos`](https://pypi.org/project/tensorflow-macos/#files) currently only supports TF 2.5.0 and Python 3.8 and 3.9.

Comment (bhack) [reply .github/workflows/release.yml:53]: And tensorflow-metal?

Comment (lgeiger) [reply .github/workflows/release.yml:53]: As far as I can tell `tensorflow-metal` is an optional dependency to enable metal support via the new pluggable device mechanism so it is not really necessary here since as far as I know TFA doesn't directly link against the metal plugin.

**Comment by lgeiger** [comment]
We briefly tested [the generated Python 3.9 wheel](https://github.com/tensorflow/addons/suites/3713514594/artifacts/90761269) and it seems to work fine on a M1 ARM Mac with `tensorflow-macos` installed.

File: .github/workflows/release.yml
Comment (bhack) [reply .github/workflows/release.yml:53]: Yes I meant if It Is constrained to 2.5.0 now I suppose that all the OSX stuffs are aligned on 2.5.0 including GPU acell

Comment (lgeiger) [reply .github/workflows/release.yml:53]: Yes, `tensorflow-metal` is also only built against 2.5.0

Comment (lgeiger) [reply .github/workflows/release.yml:53]: I am not sure how the `tensorflow-macos` wheels are generated, so there is not really a way to see whether 2.6 support is right around the corner or not.

Comment (bhack) [reply .github/workflows/release.yml:53]: What I see is that `tensorflow-macos` [is archived](https://github.com/apple/tensorflow_macos).
`tensorflow-metal` will work with official Apple conda channel [tensorflow-deps](https://anaconda.org/apple/tensorflow-deps/files) release but what it is strange is that deps has already `2.6.0`.

In the official documentation `tensorflow-deps` it is an official step also for M1:

https://developer.apple.com/metal/tensorflow-plugin/

**Review by bhack — APPROVED** [review]

**Comment by tetsuyasu** [comment]
> /cc @tetsuyasu can you check this on your system?

It works fine.

File: .github/workflows/release.yml
Comment (lgeiger) [reply .github/workflows/release.yml:53]: > What I see is that tensorflow-macos is [archived](https://github.com/apple/tensorflow_macos).

That is correct. As far as I can tell, this repo was used to [build wheels based on TF 2.4.rc0](https://github.com/apple/tensorflow_macos#supported-versions).

Step two in https://developer.apple.com/metal/tensorflow-plugin/ says to install `tensorflow-macos` which seems like it is built directly from the main https://github.com/tensorflow/tensorflow repo. Unfortunately there is not really a way to verify where the wheels are coming from.

> tensorflow-metal will work with the official Apple conda channel tensorflow-deps release but it is strange that tensorflow-deps has already 2.6.0

It looks like `tensorflow-deps` has already 2.6.0 was only released yesterday, so maybe they will update `tensorflow-macos` soon as well. As far as I know the `tensorflow-deps` package only includes the pip dependencies from TensorFlow built for ARM. Since we only need to have the prebuilt `*.so` file from TensorFlow available to link against we don't need to install `tensorflow-deps`. We cannot run it on the x86 GitHub actions VM anyway. Or am I missing something?

Comment (bhack) [reply .github/workflows/release.yml:53]: > Unfortunately there is not really a way to verify where the wheels are coming from.
Yes as  I think you already know TF CI is not as easy like to collaborate on Github Actions on the repo.

`tensorflow-macos` is still owned by Google and are the macos  ARM64 and X86_64 builds maintained by @kulinseth

Instead in the standard `tensorflow` wheel releases have ony macos X86_64 and no ARM64 coverage but `2.6.0` is available as on other platforms

> tensorflow-deps has already 2.6.0 was only released yesterday,

Yes this was just the point for guessing if `tensorflow-macos`  was imminent or not before we merge this PR but I think that only @kulinseth could help on this as he is the maintainer of both `tensorflow-macos` and `tensorflow-metal` on pypi.

If the release is imminent we could just wait and edit this PR, if not I think that we could merge and open a new one later.

Comment (lgeiger) [reply .github/workflows/release.yml:53]: > `tensorflow-macos` is still owned by Google and are the macos ARM64 and X86_64 builds maintained by @kulinseth

Oh this is interesting, I didn't know that. But then, what is the reason why the Python 3.8 and 3.9 ARM wheels are not published under the `tensorflow` in the first place? From the outside it looks like the `*macosx_11_0_arm64.whl` wouldn't conflict with any of the existing wheels.

Comment (bhack) [reply .github/workflows/release.yml:53]: Cause I think that `tensorflow` pypi macos releases are produced on the Google build infra  and SIG build ownership with:

https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/ci_build/rel/macos

Instead  `tensorflow-macos` builds that include ARM64 I think are managed by @kulinseth on external CI/Build resources

Comment (lgeiger) [reply .github/workflows/release.yml:53]: > Instead tensorflow-macos builds that include ARM64 I think are managed by @kulinseth on external CI/Build resources

I see, although I don't think PyPI requires all wheels to be uploaded at the same time, so if the external CI would have permissions to write to `tensorflow` directly this shouldn't be a problem since the ARM wheels won't conflict with any of the existing wheels already uploaded by the SIG build infra.

Comment (bhack) [reply .github/workflows/release.yml:53]: Yes, as I told you it is not about conflicts, It is about that if `tensorflow-macos` is imminent to land on 2.6.0 as `tensorflow-deps` is ready for TF 2.6.0 I will just wait to edit this PR directly to have a 2.6.0 wheel published for TF addons.
If not probably we will merge as is.

Comment (kulinseth) [reply .github/workflows/release.yml:53]: > Yes, as I told you it is not about conflicts, It is about that if `tensorflow-macos` is imminent to land on 2.6.0 as `tensorflow-deps` is ready for TF 2.6.0 I will just wait to edit this PR directly to have a 2.6.0 wheel published for TF addons.
> If not probably we will merge as is.

@mihaimaruseac and @penpornk
 
Thanks @bhack for the heads-up. Is this PR for enabling ARM64 builds for TF through official build channels which will be hosted on pypi ?

Comment (bhack) [reply .github/workflows/release.yml:53]: >  Is this PR for enabling ARM64 builds for TF through official build channels which will be hosted on pypi ?

No this is limited for Tensorflow Addons ARM64 builds in the official Tesnroflow Addons pypi so we need to know if we could enable 2.6.0 for `tensorflow-macos` 2.6.0.

> ARM64 builds for TF through official build channels which will be hosted on pypi 

For this you really need a brand new PR in the TF own repository over:
(Stable) 
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/ci_build/rel/macos
(Eventually nightly) https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/ci_build/nightly_release/macos


But it is another topic.

**Review by seanpmorgan — APPROVED** [review]
LGTM from a build standpoint. Thanks @lgeiger we've had a number of requests for this.

**Comment by lgeiger** [comment]
Great! Thanks for the fast review and merge 🚀 

How do you feel about backporting this to 0.13 and 0.14?

**Comment by bhack** [comment]
> How do you feel about backporting this to 0.13 and 0.14?

I think it is ok.

**Comment by bhack** [comment]
@lgeiger We had some problem in https://github.com/tensorflow/addons/actions/runs/1220898795

`macos-11-3.9-tf2.5.0-arm64-wheel" was not found.`

**Comment by lgeiger** [comment]
> @lgeiger We had some problem in https://github.com/tensorflow/addons/actions/runs/1220898795

Very sorry about that. #2565 should fix this typo.

**Comment by bhack** [comment]
Thanks, when it is merged can you post an [announce in the forum](https://discuss.tensorflow.org/)?

File: .github/workflows/release.yml
Comment (kulinseth) [reply .github/workflows/release.yml:53]: @bhack , thanks for the update. We are working on the v2.6 release, I will update here regarding that. I see this PR has merged we can adopt to `tensorflow-macos` after we make the release in a separate PR.

**Comment by seanpmorgan** [comment]
@bhack @lgeiger Any thoughts on how we should handle the lagging `tensorflow-macos` releases? If it's going to lag by ~month I won't be in favor of waiting for it. IMO the obvious answer is for tensorflow core to publish an arm64 macos whl... @kulinseth are there any plans to submit that PR or is a separate whl the path forward?

Other options I see:
* We could release a patch once `tensorflow-macos` catches up, but not really thrilled to absorb that burden and x.0 releases would be exempt from arm64 macos
* We could only build macos arm for tfa-nightly. And then it's just a best effort thing

...open to other ideas but would like a good reason why tf core can't publish the build along with all the others.

**Comment by lgeiger** [comment]
> IMO the obvious answer is for tensorflow core to publish an arm64 macos whl

I fully agree. To me, it seems like there shouldn't be any technical blockers for a proper arm64 macos wheel as it can be cross compiled on the existing macos CI infra. It is probably just a matter of priorities for the TF/Apple teams.

**Comment by bhack** [comment]
I think It is relatively easy to make a PR 
in the [MacOS directory I've mentioned](https://github.com/tensorflow/addons/pull/2559#discussion_r705632209).
But the problem is more related to the current CI script orchestration (Kokoro/Jenkins) that It isn't so clear for a contributor as Github actions cause it is not hosted in the repository so we don't have exactly the full overview of the CI.

The second point is:
Do we really need to wait for the `tensorflow-metal` release also on M1? 
Can `tensorflow-metal` depend on the official TF wheel instead of `tensorflow-macos` or is there something different in this build other then the arm64 support?

**Comment by lgeiger** [comment]
Sorry for the late reply, I was away the last weeks.

> Do we really need to wait for the `tensorflow-metal` release also on M1?

`tensorflow-metal` is just the device plugin for metal support for M1 Macs, so as far as I can tell `tensorflow-macos` can be used without `tensorflow-metal` just fine.

> @lgeiger Can you check the CI in the latest PRs?

@bhack Could you clarify what you mean by that?

**Comment by bhack** [comment]
> Could you clarify what you mean by that?

Nevermind,  MacOS builds were failing but it is solved.

File: .github/workflows/release.yml
Comment (bhack) [reply .github/workflows/release.yml:53]: @kulinseth Are you interested to manage the TF addons wheels release on your pypi on a new `tensorflow-macos/metal` release?

**Comment by bhack** [comment]
@seanpmorgan We could have some effect on the chain if we don't (or with @kulinseth) handle the M1 release.
See what happens in TF federated:
https://github.com/tensorflow/federated/issues/1254#issuecomment-803938932

**Comment by seanpmorgan** [comment]
> @seanpmorgan We could have some effect on the chain if we don't (or with @kulinseth) handle the M1 release. See what happens in TF federated: [tensorflow/federated#1254 (comment)](https://github.com/tensorflow/federated/issues/1254#issuecomment-803938932)

Given discussion at the last SIG meeting, we won't support TFA releases on M1 ARM until an acceptable SLA can be given from Apple on their release of tensorflow-macos. I'm okay with continuing the TFA-Nightly support though while we wait to see what happens. Only issue there is tfa-nightly arm may be compiled against an older TF version (would only affect custom ops)

**Pull request merged**
